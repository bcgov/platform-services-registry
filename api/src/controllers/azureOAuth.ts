import axios from "axios";
import { logger } from "@bcgov/common-nodejs-utils";
// import { Response } from "express";
import qs from "qs";
import { Request, Response } from "express";

export const fetchAzureAccessToken = async (req: Request, res: Response): Promise<void> => {
  try {
    const options = {
      Host: "login.microsoftonline.com",
      client_id: `${process.env.AZURE_CLIENT_ID}`,
      scope: "https://graph.microsoft.com/.default",
      client_secret: process.env.AZURE_CLIENT_SECRET,
      grant_type: "client_credentials",
    };
    const url: string = `https://login.microsoftonline.com/${process.env.AZURE_TENANT_ID}/oauth2/v2.0/token`;
    axios
      .post(url, qs.stringify(options))
      .then((response) => {
        if (!response.data.access_token) {
          throw Error(
            "a result was returned from Microsoft's token endpoint, but it doesn't seem to have an access token included"
          );
        }
        res.status(200).json(response.data.access_token);
      })
      .catch((error) => {
        logger.info(`error: ${error}`);
        res.status(500);
      });
  } catch (err) {
    logger.error(`err = ${err.message}`);
  }
};

export default fetchAzureAccessToken;
