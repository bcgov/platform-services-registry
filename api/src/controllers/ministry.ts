//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { errorWithCode, logger } from "@bcgov/common-nodejs-utils";
import { Request, Response } from "express";
import DataManager from "../db";
import shared from "../libs/shared";

const dm = new DataManager(shared.pgPool);

const fetchMinistrySponsors = async (
  req: Request,
  res: Response
): Promise<void> => {
  const { MinistryModel } = dm;
  try {
    const results = await MinistryModel.findAll();

    res.status(200).json(results);
  } catch (err) {
    const message = "Unable fetch ministries";
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export default fetchMinistrySponsors;
