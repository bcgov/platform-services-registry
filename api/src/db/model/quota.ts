//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { logger } from '@bcgov/common-nodejs-utils';
import { Pool } from 'pg';
import { CommonFields, Model } from './model';

export enum QuotaSize {
    Small = 'small',
    Medium = 'medium',
    Large = 'large',
}

export interface Quotas {
    cpu: {
        requests: number;
        limits: number;
    };
    memory: {
        requests: string;
        limits: string;
    };
    storage: {
        block: string;
        file: string;
        backup: string;
        capacity: string;
        pvcCount: number;
    };
}

export interface Quota extends CommonFields {
    cpuRequests: number;
    cpuLimits: number;
    memoryRequests: string;
    memoryLimits: string;
    storageBlock: string;
    storageFile: string;
    storageBackup: string;
    storageCapacity: string;
    storagePvcCount: number;
}

export default class QuotaModel extends Model {
    table: string = 'ref_quota';
    requiredFields: string[] = [
        'cpu_requests',
        'cpu_limits',
        'memory_requests',
        'memory_limits',
        'storage_block',
        'storage_file',
        'storage_backup',
        'storage_capacity',
        'storage_pvc_count',
    ];
    pool: Pool;

    constructor(pool: any) {
        super();
        this.pool = pool;
    }

    async create(): Promise<any> {
        // this is intentional (required by Sonarcloud)
    }

    async update(): Promise<any> {
        // this is intentional (required by Sonarcloud)
    }

    async delete(): Promise<any> {
        // this is intentional (required by Sonarcloud)
    }

    async findQuota(): Promise<any[]> {
        const query = {
            text: `
          SELECT * FROM ${this.table};
          `,
        };

        try {
            return await this.runQuery(query);
        } catch (err) {
            const message = `Unable to fetch all Quota`;
            logger.error(`${message}, err = ${err.message}`);

            throw err;
        }
    }

    async findQuotaSizes(): Promise<any> {
        try {
            const quota = await this.findQuota();
            const quotaSizes: any = [];

            for (let size of quota) {
                quotaSizes.push(
                    {
                        name: size.id,
                        cpuNums: [size.cpuRequests, size.cpuLimits],
                        memoryNums: [size.memoryRequests.replace("Gi", "GiB"), size.memoryLimits.replace("Gi", "GiB")],
                        storageNums: [size.storagePvcCount, size.storageFile.replace("Gi", "GiB"), size.storageBackup.replace("Gi", "GiB")],
                    },
                );
            }

            return quotaSizes;
        } catch (err) {
            const message = `Unable to get quota sizes`;
            logger.error(`${message}, err = ${err.message}`);

            throw err;
        }
    }

    async findForQuotaSize(quotaSize: QuotaSize): Promise<any> {
        const query = {
            text: `
                SELECT json_build_object(
                    'cpu', (SELECT row_to_json(d) FROM (SELECT cpu_requests AS "requests", cpu_limits AS "limits"
                        FROM ref_quota WHERE id = '${quotaSize}') d),
                    'memory', (SELECT row_to_json(d) FROM (SELECT memory_requests AS "requests", memory_limits AS "limits"
                        FROM ref_quota WHERE id = '${quotaSize}') d),
                    'storage', (SELECT row_to_json(d) FROM (SELECT storage_block AS "block", storage_file AS "file", storage_backup AS "backup", storage_capacity AS "capacity", storage_pvc_count AS "pvcCount"
                        FROM ref_quota WHERE id = '${quotaSize}') d)
                );
            `
        };

        try {
            const results = await this.runQuery(query);
            return results.pop().jsonBuildObject;
        } catch (err) {
            const message = `Unable to retrieve quotas object by size ${quotaSize}`;
            logger.error(`${message}, err = ${err.message}`);

            throw err;
        }
    }
}
