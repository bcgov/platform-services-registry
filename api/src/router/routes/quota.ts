//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-06-15.
//

'use strict';

import { asyncMiddleware } from '@bcgov/common-nodejs-utils';
import express from 'express';
import { fetchQuota, fetchQuotaSizes } from '../../controllers/quota';

const router = express.Router();

// Quota
// Fetches raw quota specifications from DB
// This endpoint is not being utilized by the front end; implemented for future use
router.get('/', asyncMiddleware(fetchQuota));

// Utilizes raw data from DB
// and transfroms to usable data for the front end
router.get('/sizes', asyncMiddleware(fetchQuotaSizes));

export default router;
