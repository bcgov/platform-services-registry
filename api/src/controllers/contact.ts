//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-06-15.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import shared from '../libs/shared';
import { validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const createContact = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { ContactModel } = dm;

  const rv = validateObjProps(ContactModel.requiredFields, body);
  if (rv) {
    throw rv;
  }

  try {
    const result = await ContactModel.create(body);

    res.status(201).json(result);
  } catch (err) {
    const message = `Unable to create contact`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateContact = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { ContactModel } = dm;
  const { contactId } = params;
  const {
    firstName,
    lastName,
    email,
    githubId,
  } = body;

  try {
    const record = await ContactModel.findById(contactId);
    const aBody = {
      userId: record.userId,
      firstName,
      lastName,
      email,
      githubId,
      roleId: record.roleId,
    };

    const rv = validateObjProps(ContactModel.requiredFields, aBody);

    if (rv) {
      throw rv;
    }

    const results = await ContactModel.update(contactId, aBody);

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable update contact ID ${contactId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const requestContactEdit = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  try {
    res.status(201).end();
  } catch (err) {
    const message = `Unable to update quota`;
    logger.error(`${message}, err = ${err.message}`);
    throw errorWithCode(message, 500);
  }
};
