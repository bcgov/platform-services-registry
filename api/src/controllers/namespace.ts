//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-27.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import shared from '../libs/shared';
import { validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const createNamespace = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId } = params;
  const aBody = { ...body, profileId };

  const rv = validateObjProps(NamespaceModel.requiredFields, aBody);
  if (rv) {
    throw rv;
  }

  try {
    const results = await NamespaceModel.create(aBody);

    res.status(201).json(results);
  } catch (err) {
    const message = 'Unable create new namespace';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileNamespaces = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId } = params;

  try {
    const results = await NamespaceModel.findForProfile(Number(profileId));

    res.status(200).json(results);
  } catch (err) {
    const message = `Unable fetch namespace for profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileNamespace = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { namespaceId } = params;

  try {
    const results = await NamespaceModel.findById(Number(namespaceId));

    res.status(200).json(results);
  } catch (err) {
    const message = `Unable fetch namespace with ID ${namespaceId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProfileNamespace = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId, namespaceId } = params;
  const { name, clusterId } = body;
  const aBody = { name, profileId, clusterId };

  const rv = validateObjProps(NamespaceModel.requiredFields, aBody);
  if (rv) {
    throw rv;
  }

  try {
    await NamespaceModel.update(namespaceId, aBody);

    res.status(204).end();
  } catch (err) {
    const message = `Unable update project profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const archiveProfileNamespace = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId, namespaceId } = params;

  try {
    await NamespaceModel.delete(namespaceId);

    res.status(204).end();
  } catch (err) {
    const message = `Unable to archive profile ${profileId} namespace ${namespaceId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};
