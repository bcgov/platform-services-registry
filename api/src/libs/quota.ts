//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { logger } from '@bcgov/common-nodejs-utils';
import { NameSpacesQuotaSize, ProjectQuotaSize } from '../db/model/namespace';
import { QuotaSize } from '../db/model/quota';

export const getAllowedQuotaSizes = (currentQuotaSize: ProjectQuotaSize): NameSpacesQuotaSize => {
    try {
        const quotaSizeNames = Object.values(QuotaSize);
        const allQuotaOptions = [QuotaSize.Small, QuotaSize.Medium, QuotaSize.Large];

        const availableQuotaOptions: NameSpacesQuotaSize = {
            quotaCpuSize: [],
            quotaMemorySize: [],
            quotaStorageSize: [],
        }

        for (const key in currentQuotaSize) {
            if (currentQuotaSize[key]) {
                const num: number = quotaSizeNames.indexOf(currentQuotaSize[key]);
                availableQuotaOptions[key] = allQuotaOptions.slice(
                    0, (num + 2 <= allQuotaOptions.length) ? (num + 2) : allQuotaOptions.length).filter(
                        size => size !== currentQuotaSize[key]
                    );
            }
        }

        return availableQuotaOptions
    } catch (err) {
        const message = `Unable to get a list of Allowed quota sizes`;
        logger.error(`${message}, err = ${err.message}`);

        throw err;
    }
};
