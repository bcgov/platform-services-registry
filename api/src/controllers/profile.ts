//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-27.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import { USER_ROLES } from '../constants';
import DataManager from '../db';
import { generateNamespacePrefix } from '../db/utils';
import { AuthenticatedUser } from '../libs/authmware';
import shared from '../libs/shared';
import { isNotAuthorized, validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const uniqueNamespacePrefix = async (): Promise<string | undefined> => {
  const { ProfileModel } = dm;
  const attempts = 3;
  const candidates = Array.from({ length: attempts }, () => generateNamespacePrefix())
  const promises = candidates.map(c => ProfileModel.isNamespacePrefixUnique(c));

  try {
    const results = await Promise.all(promises);
    const values = results.map((cur, idx) => {
      if (cur) {
        return candidates[idx];
      }

      return;
    }).filter(v => typeof v !== 'undefined');

    return values.pop();
  } catch (err) {
    logger.error(err.message);
    return;
  }
}

export const fetchAllProjectProfiles = async (
  { user }: { user: AuthenticatedUser }, res: Response): Promise<void> => {
  const { ProfileModel } = dm;

  try {
    let results;
    if (user.roles.includes(USER_ROLES.ADMINISTRATOR)) {
      results = await ProfileModel.findAll();
    } else {
      results = await ProfileModel.findProfilesByUserId(user.id);
    }

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable fetch all project profiles';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const record = await ProfileModel.findById(Number(profileId));

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    res.status(200).json(record);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable fetch project profile with ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const createProjectProfile = async (
  { body, user }: { body: any, user: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const data = { ...body, userId: user.id };

  // User cannot set the namespace prefix. If it exists, this
  // overwrites it with a place holder value. It will be replaced
  // with the actual value further on.
  const rv = validateObjProps(ProfileModel.requiredFields, {
    ...data,
    namespacePrefix: 'placeholder',
  });

  if (rv) {
    throw rv;
  }

  try {
    const namespacePrefix = await uniqueNamespacePrefix();

    if (!namespacePrefix) {
      throw errorWithCode(500, 'Unable to generate unique namespace prefix');
    }

    const results = await ProfileModel.create({ ...data, namespacePrefix });

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable create new project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProjectProfile = async (
  { params, body, user }: { params: any, body: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;
  const {
    name,
    description,
    categoryId,
    busOrgId,
    prioritySystem,
    notificationEmail,
    notificationSms,
    notificationMsTeams,
    paymentBambora,
    paymentPayBc,
    fileTransfer,
    fileStorage,
    geoMappingWeb,
    geoMappingLocation,
    schedulingCalendar,
    schedulingAppointments,
    idmSiteMinder,
    idmKeycloak,
    idmActiveDir,
    other,
  } = body;

  try {
    const record = await ProfileModel.findById(profileId);
    const aBody = {
      name,
      description,
      categoryId,
      busOrgId,
      prioritySystem,
      userId: record.userId,
      namespacePrefix: record.namespacePrefix,
      notificationEmail,
      notificationSms,
      notificationMsTeams,
      paymentBambora,
      paymentPayBc,
      fileTransfer,
      fileStorage,
      geoMappingWeb,
      geoMappingLocation,
      schedulingCalendar,
      schedulingAppointments,
      idmSiteMinder,
      idmKeycloak,
      idmActiveDir,
      other,
    };

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    const rv = validateObjProps(ProfileModel.requiredFields, aBody);

    if (rv) {
      throw rv;
    }

    const results = await ProfileModel.update(profileId, aBody);

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable update project profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const archiveProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const record = await ProfileModel.findById(profileId);

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    await ProfileModel.delete(profileId);

    res.status(204).end();
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = 'Unable to archive project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const addContactToProfile = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId, contactId } = params;

  try {
    await ProfileModel.addContactToProfile(Number(profileId), Number(contactId));

    res.status(202).end();
  } catch (err) {
    const message = `Unable to create contact`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileContacts = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ContactModel, ProfileModel } = dm;
  const { profileId } = params;

  try {
    const record = await ProfileModel.findById(Number(profileId));

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    const results = await ContactModel.findForProject(Number(profileId));

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable fetch profile contacts with profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileEditRequests = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  try {
    res.status(201).end();
  } catch (err) {
    const message = `Unable to update quota`;
    logger.error(`${message}, err = ${err.message}`);
    throw errorWithCode(message, 500);
  }
};
