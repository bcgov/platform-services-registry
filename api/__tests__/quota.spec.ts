//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import fs from 'fs';
import { camelCase } from 'lodash';
import path from 'path';
import { Pool } from 'pg';
import { fetchQuota, fetchQuotaSizes } from '../src/controllers/quota';
import FauxExpress from './src/fauxexpress';

const p0 = path.join(__dirname, 'fixtures/select-quota.json');
const selectQuota = JSON.parse(fs.readFileSync(p0, 'utf8'));


const client = new Pool().connect();

jest.mock('../src/db/utils', () => ({
    generateNamespacePrefix: jest.fn().mockReturnValue('c8c7e6'),
    transformKeysToCamelCase: jest.fn().mockImplementation(data => {
        const obj = {};
        Object.keys(data).forEach(key => {
            obj[camelCase(key)] = data[key];
        });

        return obj;
    }),
}));

describe('Quota event handlers', () => {
    let ex;
    beforeEach(() => {
        jest.clearAllMocks();

        ex = new FauxExpress();
    });

    describe('fetchQuota', () => {
        it('All quotas are returned', async () => {
            const req = {};
            client.query.mockReturnValueOnce({ rows: selectQuota });

            // @ts-ignore
            await fetchQuota(req, ex.res);

            expect(client.query.mock.calls).toMatchSnapshot();
            expect(ex.res.statusCode).toMatchSnapshot();
            expect(ex.responseData).toMatchSnapshot();
            expect(ex.res.status).toBeCalledWith(ex.res.statusCode);
            expect(ex.res.json).toBeCalledWith(ex.responseData);
        });

        it('Quota query error should throw', async () => {
            const req = {};
            client.query.mockImplementation(() => { throw new Error() });

            // @ts-ignore
            await expect(fetchQuota(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();

            expect(client.query.mock.calls).toMatchSnapshot();
            expect(ex.responseData).toBeUndefined();
        });
    });

    describe('fetchQuotaSizes', () => {
        it('All quota sizes are returned', async () => {
            const req = {};
            client.query.mockReturnValueOnce({ rows: selectQuota });

            // @ts-ignore
            await fetchQuotaSizes(req, ex.res);

            expect(client.query.mock.calls).toMatchSnapshot();
            expect(ex.res.statusCode).toMatchSnapshot();
            expect(ex.responseData).toMatchSnapshot();
            expect(ex.res.status).toBeCalled();
            expect(ex.res.json).toBeCalled();
        });

        it('Quota sizes query error should throw', async () => {
            const req = {};

            client.query.mockImplementation(() => { throw new Error() });

            // @ts-ignore
            await expect(fetchQuotaSizes(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();

            expect(client.query.mock.calls).toMatchSnapshot();
            expect(ex.responseData).toBeUndefined();
        });
    });
});

