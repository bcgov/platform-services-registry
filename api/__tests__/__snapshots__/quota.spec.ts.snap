// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quota event handlers fetchQuota All quotas are returned 1`] = `
Array [
  Array [
    Object {
      "text": "
          SELECT * FROM ref_quota;
          ",
    },
  ],
]
`;

exports[`Quota event handlers fetchQuota All quotas are returned 2`] = `200`;

exports[`Quota event handlers fetchQuota All quotas are returned 3`] = `
Array [
  Object {
    "cpuLimits": 8,
    "cpuRequests": 4,
    "createdAt": "2021-03-17T23:06:05.112Z",
    "id": "small",
    "memoryLimits": "32Gi",
    "memoryRequests": "16Gi",
    "storageBackup": "25Gi",
    "storageBlock": "50Gi",
    "storageCapacity": "50Gi",
    "storageFile": "50Gi",
    "storagePvcCount": 20,
    "updatedAt": "2021-03-17T23:06:05.112Z",
  },
  Object {
    "cpuLimits": 16,
    "cpuRequests": 8,
    "createdAt": "2021-03-17T23:06:05.112Z",
    "id": "medium",
    "memoryLimits": "64Gi",
    "memoryRequests": "32Gi",
    "storageBackup": "50Gi",
    "storageBlock": "100Gi",
    "storageCapacity": "100Gi",
    "storageFile": "100Gi",
    "storagePvcCount": 40,
    "updatedAt": "2021-03-17T23:06:05.112Z",
  },
  Object {
    "cpuLimits": 64,
    "cpuRequests": 32,
    "createdAt": "2021-03-17T23:06:05.112Z",
    "id": "large",
    "memoryLimits": "128Gi",
    "memoryRequests": "64Gi",
    "storageBackup": "100Gi",
    "storageBlock": "200Gi",
    "storageCapacity": "200Gi",
    "storageFile": "200Gi",
    "storagePvcCount": 60,
    "updatedAt": "2021-03-17T23:06:05.112Z",
  },
]
`;

exports[`Quota event handlers fetchQuota Quota query error should throw 1`] = `"Unable fetch quotas"`;

exports[`Quota event handlers fetchQuota Quota query error should throw 2`] = `
Array [
  Array [
    Object {
      "text": "
          SELECT * FROM ref_quota;
          ",
    },
  ],
]
`;

exports[`Quota event handlers fetchQuotaSizes All quota sizes are returned 1`] = `
Array [
  Array [
    Object {
      "text": "
          SELECT * FROM ref_quota;
          ",
    },
  ],
]
`;

exports[`Quota event handlers fetchQuotaSizes All quota sizes are returned 2`] = `200`;

exports[`Quota event handlers fetchQuotaSizes All quota sizes are returned 3`] = `
Array [
  Object {
    "cpuNums": Array [
      4,
      8,
    ],
    "memoryNums": Array [
      "16GiB",
      "32GiB",
    ],
    "name": "small",
    "storageNums": Array [
      20,
      "50GiB",
      "25GiB",
    ],
  },
  Object {
    "cpuNums": Array [
      8,
      16,
    ],
    "memoryNums": Array [
      "32GiB",
      "64GiB",
    ],
    "name": "medium",
    "storageNums": Array [
      40,
      "100GiB",
      "50GiB",
    ],
  },
  Object {
    "cpuNums": Array [
      32,
      64,
    ],
    "memoryNums": Array [
      "64GiB",
      "128GiB",
    ],
    "name": "large",
    "storageNums": Array [
      60,
      "200GiB",
      "100GiB",
    ],
  },
]
`;

exports[`Quota event handlers fetchQuotaSizes Quota sizes query error should throw 1`] = `"Unable to get quota sizes"`;

exports[`Quota event handlers fetchQuotaSizes Quota sizes query error should throw 2`] = `
Array [
  Array [
    Object {
      "text": "
          SELECT * FROM ref_quota;
          ",
    },
  ],
]
`;
